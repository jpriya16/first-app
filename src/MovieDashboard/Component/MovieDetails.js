import React from "react";
export default function MovieDetails(Movies){
    return(
        <div>
            <table>
                <tr>
                    <th>Movie name</th>
                    <th>ID</th>
                    <th>Total tickets</th>
                    <th>Booked</th>
                </tr>
                { Movies.map( (val, key) => {
                    return (
                        <tr key={key}>
                            <td>{val.name}</td>
                            <td>{val._id}</td>
                            <td>{val.total_tickets}</td>
                            <td>{val.booked_tickets}</td>
                        </tr>
                    )
                }) }
            </table>
        </div>
    )
}
