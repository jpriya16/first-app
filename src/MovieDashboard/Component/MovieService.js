import axios from "axios";


    export const getMovies = () => {

        return axios.get(`https://crudcrud.com/api/29bd8cc907024d05b532b554ddf9f979/movies`)
            .then(res => {
                return res.data
            })
            .catch((error) => {return []} )
    }

    export const updateTicket = (movieId,Movie) => {
        return axios.put(`https://crudcrud.com/api/29bd8cc907024d05b532b554ddf9f979/movies/` + movieId, Movie)
            .then(response => {
               return true
            })
            .catch((error)=>{
                return false
            })
     }


