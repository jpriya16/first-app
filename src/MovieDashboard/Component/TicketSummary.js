import React from "react";

export default function TicketSummary (movies) {
    let totalTickets, bookedTickets;
    totalTickets = 0;
    bookedTickets = 0;
    let availableTickets = 0;

    for(let movie in movies){
            totalTickets += Number(movies[movie].total_tickets);
            bookedTickets += Number(movies[movie].booked_tickets);
    }
    availableTickets = totalTickets - bookedTickets;

    return(<div>
        <table>
            <tr><th id="total"> Total </th></tr>
            <tr><td id="availableTicket">Available Tickets: {availableTickets}</td></tr>
            <tr><td id="bookedTicket">Booked Count: { bookedTickets }</td></tr>
        </table>
    </div>)
}
