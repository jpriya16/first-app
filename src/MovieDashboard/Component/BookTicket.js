import React from "react";
import {useState} from "react";
import {updateTicket} from "./MovieService";

export default function BookTicket(movies,UpdateMovieDetails){
    const [MovieId, setMovieId] = useState('')
    const [NumberOfTickets,setNumberOfTickets] = useState(0)
    const [Book,setBook] = useState(false)
    const [message,setMessage] = useState('')


    const handleBooKTicket = () => {
        setBook(true);
        BookTicketForMovie(movies);
        setNumberOfTickets(0);
        setMovieId('');
        setBook(false)

    }

    const bookTicketWithData = async (MovieId,newMovie) => {
        return await updateTicket(MovieId,newMovie)
    }

    const BookTicketForMovie = (movies) => {
        let movie = null;
        let diff = 0;
        let total = 0;
        let booked = 0;
        for (let key in movies) {
            if(movies[key]){
                if(movies[key]._id === MovieId) {
                    console.log("movie present", movies[key])
                    movie = movies[key];
                    total = Number(movie.total_tickets);
                    booked = Number(movie.booked_tickets);
                    diff = total - (NumberOfTickets + booked)
                }
            }
        }



        if(movie == null)
            setMessage('Invalid Movie Id ' + MovieId)
        else if(diff < 0){
            setMessage('requested no of ticket is not available for Movie Id ' + MovieId)}
        else if (movie && diff >= 0) {
                let newMovie = {
                    "name": movie.name,
                    "total_tickets": total,
                    "booked_tickets": booked + NumberOfTickets
                }
                bookTicketWithData(MovieId, newMovie).then(r =>{
                    if (r){
                        UpdateMovieDetails()
                        setMessage('Tickets booked for Movie Id : '+ MovieId +'Successfully')}
                }).catch(e=>{

                        setMessage('try booking after some time')
                })
                // axios.put(`https://crudcrud.com/api/c52d5cdcc84e415e80d8f2b006cddc0b/movies/` + MovieId, newMovie).then(response => {
                //     UpdateMovieDetails()
                //     setMessage('Tickets booked for Movie Id : '+ MovieId +'Successfully')
                //     }
                // ).catch(error => {
                //     setMessage('try booking after some time')
                // })


        }
    }

    return(
        <div>
            <label>MOVIE DASHBOARD</label>
            <br/><br/>
            <label>MOVIE ID</label>
            <input data-testid ="movieId" id ="movieId" placeholder="Movie Id" value={MovieId} onChange={e=>setMovieId(e.target.value)}></input>
            <label>Number of tickets to book</label>
            <input data-testid="noOfTicket" id="noOfTicket" type="text" pattern="[0-9]*" value={NumberOfTickets} placeholder="no of tickets" onChange={e=>setNumberOfTickets(Number(e.target.value))}></input>
            <button data-testid="bookButton" id ="bookbtn" onClick={handleBooKTicket}>Book</button>
            { Book || <label>{message}</label>}
        </div>
    )
}
