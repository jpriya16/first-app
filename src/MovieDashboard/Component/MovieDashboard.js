import React from "react";
import axios from "axios";
import {useState} from "react";
import {useEffect} from "react";
import TicketSummary from "./TicketSummary";
import MovieDetails from "./MovieDetails";
import BookTicket from "./BookTicket";
import {getMovies} from "./MovieService";

export default function MovieDashboard(){
    const [Movies,setMovies] = useState([])

    useEffect(async () => {
        setMovies(await getMovies());
    },[])

    const updateMovieDetails = async () => {
        let movieData = await getMovies();
        setMovies(movieData)
    }


    return(<div>
        {BookTicket(Movies,updateMovieDetails)}
        <br/>
        <br/>
        {MovieDetails(Movies)}
        <br/>
        {TicketSummary(Movies)}
    </div>);


}
