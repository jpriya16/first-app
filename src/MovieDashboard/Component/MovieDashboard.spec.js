import {configure, mount, shallow} from "enzyme";
import TicketSummary from "./TicketSummary";
import MovieDetails from "./MovieDetails";
import BookTicket from "./BookTicket";
import React from "react";
import MovieDashboard from "./MovieDashboard";
import Adapter from "enzyme-adapter-react-16";
import {getMovies} from "./MovieService";
import {render} from "@testing-library/react";
configure({ adapter: new Adapter() });
describe("Movie dashboard",()=>{
    const movies =
        [
            {
                "_id": "123",
                "name": "Movie5",
                "total_tickets": 40,
                "booked_tickets": 10
            }
        ]


    it("should match snapshot",()=>{
        const component = shallow(<MovieDashboard />)
        expect(component.getElements()).toMatchSnapshot();
    })

    jest.mock('./BookTicket', () => ({
        __esModule: true,
        namedExport: jest.fn(),
        default: jest.fn()
    }));

    it("should display the Movie dashboard,movie details,total", () => {
       //  const BookTicketMock = jest.mock('./BookTicket', () => ({
       //      __esModule: true,
       //      default: jest.fn()
       //   }));
       //
       //  jest.mock('axios')
       //
       //  const getMovies = jest.fn(() => new Promise((resolve) => {
       //      resolve({data: movies})
       //  }));
       //  const TicketSummary = jest.fn(()=>{
       //  })
       //  const MovieDetails = jest.mock('./MovieDetails')
       //  //const BookTicket = jest.mock('./BookTicket')
       //  render(<MovieDashboard />)
       //  expect(BookTicket(movies,jest.fn())).toHaveBeenCalledTimes(1)
       //  //expect(<TicketSummary movies={movies}/>).toHaveBeenCalledTimes(1)
       // // expect(MovieDetails).toHaveBeenCalledTimes(1)
    })

})
