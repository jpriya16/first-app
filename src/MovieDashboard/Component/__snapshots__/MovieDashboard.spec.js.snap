// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie dashboard should match snapshot 1`] = `
Array [
  <div>
    <div>
      <label>
        MOVIE DASHBOARD
      </label>
      <br />
      <br />
      <label>
        MOVIE ID
      </label>
      <input
        data-testid="movieId"
        onChange={[Function]}
        placeholder="Movie Id"
        value=""
      />
      <label>
        Number of tickets to book
      </label>
      <input
        data-testid="noOfTicket"
        onChange={[Function]}
        pattern="[0-9]*"
        placeholder="no of tickets"
        type="text"
        value={0}
      />
      <button
        data-testid="bookButton"
        id="bookbtn"
        onClick={[Function]}
      >
        Book
      </button>
      <label>
        
      </label>
    </div>
    <br />
    <br />
    <div>
      <table>
        <tr>
          <th>
            Movie name
          </th>
          <th>
            ID
          </th>
          <th>
            Total tickets
          </th>
          <th>
            Booked
          </th>
        </tr>
      </table>
    </div>
    <br />
    <div>
      <table>
        <tr>
          <th
            id="total"
          >
             Total 
          </th>
        </tr>
        <tr>
          <td
            id="availableTicket"
          >
            Available Tickets: 
            0
          </td>
        </tr>
        <tr>
          <td
            id="bookedTicket"
          >
            Booked Count: 
            0
          </td>
        </tr>
      </table>
    </div>
  </div>,
]
`;
